### PR TITLE
OCPBUGS-37773: Added checks to prevent vCenters from changing after install past 1 vCenter

### PR DESCRIPTION
**Pull request by vr4manta**
[OCPBUGS-37773](https://issues.redhat.com/browse/OCPBUGS-37773)

### Changes
- Updated infrastructure CRD to have a check for vSphere vCenters field to make sure quantity doesn't change (except for when adding first vCenter after upgrading to newer release)

### Notes
The changes in this PR are intended to help prevent having the number of vCenters configured change after cluster is installed.  Currently OCP is not going to support updating the number of vCenters as a Day 2 configuration.  This will be add in the future, but for now we are going to only allow adding 1 vCenter both as a Day 0 and Day 2 operation.  Adding 1 vCenter as Day 2 is allowed to allow older existing clusters, that were installed before the infrastructure object was populated, to be configured for 1 or more failure domains in the existing vSphere it was installed into.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -176,7 +176,7 @@ endif
 
 .PHONY: images
 images:
-	$(ENGINE) build -f Dockerfile.rhel8 -t "$(IMAGE):$(VERSION)" -t "$(IMAGE):$(MUTABLE_TAG)" ./
+	$(ENGINE) build -f Dockerfile.ocp -t "$(IMAGE):$(VERSION)" -t "$(IMAGE):$(MUTABLE_TAG)" ./
```
Comment (JoelSpeed) [inline Makefile:179]: Huh? Seems unrelated?

File: config/v1/tests/infrastructures.config.openshift.io/VSphereMultiVCenters.yaml [hunk 2]
```diff
@@ -85,8 +85,18 @@ tests:
                 port: 443
                 server: server2.dev.cluster.com
   onUpdate:
-    - name: Should be able to set vCenters to multiple
+    - name: Should be able to set vCenters to single from empty spec
```
Comment (JoelSpeed) [inline config/v1/tests/infrastructures.config.openshift.io/VSphereMultiVCenters.yaml:88]: What's the use case here?

File: config/v1/types_infrastructure.go [hunk 3]
```diff
@@ -1340,13 +1340,14 @@ type VSpherePlatformNodeNetworking struct {
...
 	// vcenters holds the connection details for services to communicate with vCenter.
-	// Currently, only a single vCenter is supported.
+	// Currently, only a single vCenter is supported, but in tech preview 3 vCenters are supported.
 	// ---
 	// + If VCenters is not defined use the existing cloud-config configmap defined
 	// + in openshift-config.
 	// +kubebuilder:validation:MinItems=0
 	// +openshift:validation:FeatureGateAwareMaxItems:featureGate="",maxItems=1
 	// +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,maxItems=3
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf || size(self) == size(oldSelf) || (size(self) == 1 && size(oldSelf) == 0)",message="vCenters cannot be added or removed once set"
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:1350]: Does the first clause here matter? Do we care if they are identical or can we just validate the size changes?

File: config/v1/types_infrastructure.go [hunk 4]
```diff
@@ -1340,13 +1340,14 @@ type VSpherePlatformNodeNetworking struct {
 // +kubebuilder:validation:XValidation:rule="!has(oldSelf.ingressIPs) || has(self.ingressIPs)",message="ingressIPs list is required once set"
 type VSpherePlatformSpec struct {
 	// vcenters holds the connection details for services to communicate with vCenter.
-	// Currently, only a single vCenter is supported.
+	// Currently, only a single vCenter is supported, but in tech preview 3 vCenters are supported.
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:1343]: We should explain that they cannot be added/remove/changed once configured, so that the validation you're adding is expected

File: Makefile (hunk 1 above)
Comment (vr4manta) [reply Makefile:179]: Sorry, i'll remove this.  I'll put it in a new PR.  Someone changed the name of the dockerfile, but never updated the make file which I use to do my builds.

File: config/v1/tests/infrastructures.config.openshift.io/VSphereMultiVCenters.yaml (hunk 2 above)
Comment (vr4manta) [reply config/v1/tests/infrastructures.config.openshift.io/VSphereMultiVCenters.yaml]: This is in the case user upgrades cluster from a version that did not have vsphere platform spec and wanted to create failure domains.  In this case, you will need to set a vCenter in the vcenters section.

File: config/v1/types_infrastructure.go (hunk 3 above)
Comment (vr4manta) [reply config/v1/types_infrastructure.go]: sorry, this was an older push.  I've been stuck testing a scenario where the vsphere section did not exist.  The new push should have the adjusted rule.

File: config/v1/types_infrastructure.go [hunk 5]
```diff
@@ -1338,15 +1339,20 @@ type VSpherePlatformNodeNetworking struct {
...
 // +kubebuilder:validation:XValidation:rule="!has(oldSelf.apiServerInternalIPs) || has(self.apiServerInternalIPs)",message="apiServerInternalIPs list is required once set"
 // +kubebuilder:validation:XValidation:rule="!has(oldSelf.ingressIPs) || has(self.ingressIPs)",message="ingressIPs list is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.vcenters) && has(self.vcenters) ? size(self.vcenters) < 2 : true",message="vcenters count can only be one when performing day 2 config"
 type VSpherePlatformSpec struct {
 	// vcenters holds the connection details for services to communicate with vCenter.
-	// Currently, only a single vCenter is supported.
+	// Currently, only a single vCenter is supported, but in tech preview 3 vCenters are supported.
+	// Once the cluster has been installed, you are unable to change the current number of defined
+	// vCenters except in the case where the cluster has been upgraded from a version of OpenShift
+	// where the vsphere platform spec was not present.
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:1348]: "In which case, you may configure the existing failure vCenter details" or something to that effect to finish this thought?

File: config/v1/types_infrastructure.go [hunk 6]
```diff
@@ -269,6 +269,7 @@ type ExternalPlatformSpec struct {
 // PlatformSpec holds the desired state specific to the underlying infrastructure provider
 // of the current cluster. Since these are used at spec-level for the underlying cluster, it
 // is supposed that only one of the spec structs is set.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.vsphere) && has(self.vsphere) ? size(self.vsphere.vcenters) < 2 : true",message="vsphere can only have 1 vCenter when performing day 2 config"
```
[the remark this answers is not in the document]
Comment (vr4manta) [reply config/v1/types_infrastructure.go:272]: Sure, i'll update that text.

File: config/v1/types_infrastructure.go (hunk 5 above)
Comment (vr4manta) [reply config/v1/types_infrastructure.go:1348]: ok, i'll add those details.  Thanks!

**Review by JoelSpeed — COMMENTED** [review]
/lgtm
